cpp/src_prims/matrix/batched_matrix.h

```diff
@@ -0,0 +1,465 @@
+    // std::cout << "Allocate (" << shape.first << "," << shape.second << "):" << A_array << "\n";
+    in_use = true;
+  }
+
+  double* A_dense;
+  double** A_array;
+  bool in_use;
+};
+
+class BatchedMatrixMemoryPool {
+ public:
+  BatchedMatrixMemoryPool(int num_batches) : m_num_batches(num_batches) {
+    // std::cout << "Memory Pool Init\n";
+    CUBLAS_CHECK(cublasCreate(&m_handle));
+  }
+
+  ~BatchedMatrixMemoryPool() {
+    for (auto& kv : m_pool) {
+      for (auto& v : kv.second) {
+        // std::cout << "cudaFree Deallocating\n";
+        cudaFreeT(v.A_array);
+        cudaFreeT(v.A_dense);
+      }
+    }
+    CUBLAS_CHECK(cublasDestroy(m_handle));
+  }
+
+  std::pair<double*, double**> get(std::pair<int, int> shape,
+                                   bool setZero = false) {
+    auto mempool_for_shape = m_pool.find(shape);
+    // if mempool empty for this shape, create first entry
+    if (mempool_for_shape == m_pool.end()) {
+      BatchedMatrixMemory mem(shape, m_num_batches, setZero);
+      m_pool[shape] = {mem};
+      return std::make_pair(mem.A_dense, mem.A_array);
+    } else {
+      // Look for an empty (in_use==false) slot for this shape.
+      auto& shape_pool = m_pool[shape];
+      bool found_ununsed_entry = false;
+      for (auto& m : shape_pool) {
+        if (m.in_use == false) {
+          // std::cout << "Re-Use("  << shape.first << "," << shape.second << "):" << m.A_array << "\n";
+          found_ununsed_entry = true;
+          m.in_use = true;
+          if (setZero) {
+            CUDA_CHECK(cudaMemset(m.A_dense, 0.0,
+                                  shape.first * shape.second * m_num_batches));
+          }
+          return std::make_pair(m.A_dense, m.A_array);
+        }
+      }
+      if (!found_ununsed_entry) {
+        // Create a new memory slot because no entries are free.
+        BatchedMatrixMemory mem(shape, m_num_batches, setZero);
+        shape_pool.push_back(mem);
+        return std::make_pair(mem.A_dense, mem.A_array);
+      }
+    }
+    throw std::runtime_error("ERROR: BMMP::Unreachable place!");
+  }
+
+  // Free the entry with this shape and pointer for re-use.
+  void remove(std::pair<int, int> shape, double** A_array) {
+    auto& shape_pool = m_pool.at(shape);
+    bool found_array = false;
+    for (auto& m : shape_pool) {
+      if (m.A_array == A_array) {
+        found_array = true;
+        m.in_use = false;
+        break;
+      }
+    }
+    if (!found_array) {
+      std::cout << "ERROR, Couldnt find:(" << shape.first << "," << shape.second
+                << "): " << A_array << ")\n";
+      throw std::runtime_error("ERROR: Tried to remove an array not in pool");
+    }
+  }
+
+  const cublasHandle_t& cublasHandle() const { return m_handle; }
+
+ private:
+  std::unordered_map<std::pair<int, int>, std::vector<BatchedMatrixMemory>,
+                     pair_hash>
+    m_pool;
+  int m_num_batches;
+  cublasHandle_t m_handle;
+};
+
+class BatchedMatrix {
```

Comment:
Can you make this class general by templating it on the type? Currently, it appears that this class only support double type.